include/engine/error.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef OSRM_ENGINE_ERROR_HPP
+#define OSRM_ENGINE_ERROR_HPP
+
+#include "util/exception.hpp"
+
+#include <string>
+#include <array>
+
+namespace osrm
+{
+namespace engine
+{
+
+enum class ErrorCode
+{
+    NO_ERROR = 0,
+    NOT_IMPLEMENTED,
+    INVALID_VALUE,
+    INVALID_OPTIONS,
+    TOO_BIG,
+    NO_ROUTE,
+    NO_SEGMENT,
+    NO_TABLE,
+    NO_TRIP,
+    NO_MATCH,
+    NUM_ERROR,
+};
+
+inline std::string codeToString(ErrorCode code)
+{
+    static std::array<const char *, static_cast<std::size_t>(ErrorCode::NUM_ERROR)> names{
+        {"Ok",
+         "NotImplemented",
+         "InvalidValue",
+         "InvalidOptions",
+         "TooBig",
+         "NoRoute",
+         "NoSegment",
+         "NoTable",
+         "NoTrip",
+         "NoMatch"}};
+
+    return names[static_cast<std::size_t>(code)];
+}
+
+struct Error
+{
+    ErrorCode code = ErrorCode::NO_ERROR;
+    std::string message;
+
+    auto throwException() const
+    {
+        throw util::exception(std::string(codeToString(code)) + ":" + message);
+    }
+};
+
+template<typename ResultT>
+struct MaybeResult
```

Comment:
Should the error type be a template and then should the generic Error type above be split e.g. into

    TripError
    MatchError

etc.? As is now we can assign semantically meaningless error codes to plugin responses (e.g. NoTrip to the MatchPlugin response).